### PR TITLE
[NETBEANS-4938] Fixed Gradle Java Frontend Application generation

### PR DESCRIPTION
**Pull request by lkishalmi**
Well, I've made the template copy straight. It loads all three projects (root, web, desktop) also the Demo.java into the editor.

### CODE REVIEW

File: java/gradle.htmlui/src/org/netbeans/modules/gradle/htmlui/GradleArchetype.java [hunk 1]
```diff
@@ -18,59 +18,74 @@
...
-import org.netbeans.api.templates.FileBuilder;
-import org.netbeans.api.templates.FileBuilder.Mode;
+import org.netbeans.modules.gradle.spi.newproject.TemplateOperation;
 import org.openide.filesystems.FileObject;
 import org.openide.filesystems.FileUtil;
 import org.openide.util.MapFormat;
 
 public final class GradleArchetype {
     private final FileObject templates;
-    private final FileObject projectFo;
```
Comment (jtulach) [inline java/gradle.htmlui/src/org/netbeans/modules/gradle/htmlui/GradleArchetype.java:33]: I always prefer the `FileObject` abstraction over using plain `File`. It makes me sad when I see code moving to the opposite direction.

What is the reason for making such change? The fact that modifications using `File` aren't notified as soon as they are made to various listeners?

File: java/gradle.htmlui/src/org/netbeans/modules/gradle/htmlui/GradleArchetype.java [hunk 2]
```diff
@@ -18,59 +18,74 @@
...
-            String relativeParent = FileUtil.getRelativePath(templates, template.getParent());
-            Object packageAttr = template.getAttribute("package"); // NOI18N
-            if (packageAttr instanceof String) {
-                String packageName = mf.format(packageAttr).replace('.', '/');
-                relativeParent += "/" + packageName;
-            }
-            FileObject destinationFolder = FileUtil.createFolder(projectFo, relativeParent);
+            String relativePath = FileUtil.getRelativePath(templates, template);
+            if (template.isFolder()) {
+                File dir = new File(rootDir, relativePath);
```
Comment (jtulach) [inline java/gradle.htmlui/src/org/netbeans/modules/gradle/htmlui/GradleArchetype.java:55]: Relative path is using `/` as a directory separator. `File` is supposed to be using `File.separator`.

File: java/gradle.htmlui/src/org/netbeans/modules/gradle/htmlui/GradleArchetype.java [hunk 3]
```diff
@@ -18,59 +18,74 @@
...
 
-            FileBuilder fb = new FileBuilder(template, destinationFolder);
-            fb.withParameters(params);
-            fb.defaultMode(Mode.COPY);
-
-            FileObject copied = fb.build().iterator().next();
-
-            assert copied != null && copied.getNameExt().equals(template.getNameExt()) : "Created " + copied;
+    private static void copyDataTemplate(TemplateOperation ops, FileObject template, File target, Map<String, Object> params) {
+        Object importantAttr = template.getAttribute("important");
```
Comment (jtulach) [inline java/gradle.htmlui/src/org/netbeans/modules/gradle/htmlui/GradleArchetype.java:84]: The `important` attribute enhances the API, but I am not sure if the previous API was documented anywhere.

File: java/gradle.htmlui/src/org/netbeans/modules/gradle/htmlui/HtmlJavaApplicationProjectWizard.java [hunk 4]
```diff
@@ -57,47 +53,24 @@ public HtmlJavaApplicationProjectWizard() {
...
 
         FileObject folder = ((TemplateWizard)this.getData()).getTemplate().getPrimaryFile();
-        ops.addConfigureProject(projectDir, new CopyTree(folder, projectDir, mainParams));
-    }
-
-    private static class CopyTree implements TemplateOperation.ProjectConfigurator {
-        private final FileObject templateFolder;
-        private final File projectDir;
-        private final Map<String, Object> params;
+        GradleArchetype ga = new GradleArchetype(folder, rootDir, params);
```
Comment (jtulach) [inline java/gradle.htmlui/src/org/netbeans/modules/gradle/htmlui/HtmlJavaApplicationProjectWizard.java:66]: This certainly looks like a simplification.

**Review by jtulach — APPROVED** [review]
The `collectOperations` logic has been quite simplified and that is probably the most important benefit this change provides.

File: java/gradle.htmlui/src/org/netbeans/modules/gradle/htmlui/GradleArchetype.java (hunk 1 above)
Comment (lkishalmi) [reply java/gradle.htmlui/src/org/netbeans/modules/gradle/htmlui/GradleArchetype.java:33]: The reason using File object during the template steps, as they are non-existent ones while the operation steps are collected, so in that case FileObjects would be null.
The templating has two major phase: collecting the steps to be executed and then execute the steps.

File: java/gradle.htmlui/src/org/netbeans/modules/gradle/htmlui/GradleArchetype.java (hunk 2 above)
Comment (lkishalmi) [reply java/gradle.htmlui/src/org/netbeans/modules/gradle/htmlui/GradleArchetype.java:55]: Well, that's true for the correctness, though ```/``` as separator works on every platform.

File: java/gradle.htmlui/src/org/netbeans/modules/gradle/htmlui/GradleArchetype.java (hunk 3 above)
Comment (lkishalmi) [reply java/gradle.htmlui/src/org/netbeans/modules/gradle/htmlui/GradleArchetype.java:84]: Yes that's true, though this module is not providing any public API/SPI at the moment. Maybe some next iterations when the GradleArtifact would be made available for reuse. Infortunately I have at least two branches open to improve the Gradle Wizards, but I have not got the chance to bring them in a presentable state.

File: java/gradle.htmlui/src/org/netbeans/modules/gradle/htmlui/GradleArchetype.java (hunk 2 above)
Comment (jtulach) [reply java/gradle.htmlui/src/org/netbeans/modules/gradle/htmlui/GradleArchetype.java:55]: I know that OpenVMS isn't using `/` as separator. Long time ago we even got a bug report from someone using NetBeans on such system. But let it be...

File: java/gradle.htmlui/src/org/netbeans/modules/gradle/htmlui/GradleArchetype.java (hunk 3 above)
Comment (jtulach) [reply java/gradle.htmlui/src/org/netbeans/modules/gradle/htmlui/GradleArchetype.java:84]: When you think of improving the wizard, there are two things I'd suggest to address the infamous Gradle problems:
* fix the Gradle version - e.g. always generate `gradlew` - Gradle isn't trying to keep any backward compatibility. The only way to create a Gradle project and make it work after one, two, three years is to fix Gradle version with `gradlew`.
* the wizard should select (a recommended) JDK - Gradle is well known to not work with newest JDKs - the wizard should push user strongly to provide a JDK which is known to work for the project and/or selected Gradle version

Unrelated to this issue, of course. Just my 2 Kč experience of making Geertjan's [java frontend application](https://github.com/JaroslavTulach/oraclejet4j) working again after twelve (only!) months.

Comment (lkishalmi) [reply java/gradle.htmlui/src/org/netbeans/modules/gradle/htmlui/GradleArchetype.java:84]: The JDK selection is already on my branch.
I can set the wrapper on by default right now, that would be a trivial change and improve the compatibility especially as Gradle 7.0 is already on the design desk.
It seems I have to fix the tests as well.

**Comment by JaroslavTulach** [comment]
Great, it works: 
obrazek

**Pull request merged**